### ISSUE TITLE
[Feature] support pass data from Terraform cloud resource to helm component

### ISSUE THREAD

**Issue by wonderflow** [kubevela/kubevela#4339]
**Is your feature request related to a problem? Please describe.**

Below is my application that the RDS component will generate a secret name `db-conn`, it has a  `DB_PUBLIC_HOST` key with the valueI want to pass to helm component.

```
apiVersion: core.oam.dev/v1beta1
kind: Application
metadata:
  name: webapp
spec:
  components:
    - name: sample-db
      type: alibaba-rds
      properties:
        instance_name: sample-db
        account_name: oamtest
        password: U34rfwefwefffaked
        writeConnectionSecretToRef:
          name: db-conn
    - name: app
      type: helm
      properties:
        repoType: "helm"
        url: "xxx"
        chart: "app"
        version: "1.0.0"
        values:
          datasource:
            host: # how th get the secret key-value ("DB_PUBLIC_HOST") from rds db-conn?
```

**Describe the solution you'd like**

Add a velaql into component definition.

```
apiVersion: core.oam.dev/v1beta1
kind: ComponentDefinition
metadata:
  annotations:
    definition.oam.dev/description: Describes long-running, scalable, containerized services that have a stable network endpoint to receive external network traffic from customers.
  name: webservice
  namespace: {{ include "systemDefinitionNamespace" . }}
spec:
  schematic:
    cue:
      template: |
        ...
  status:
    velaql: |-
         import "velaql/op"
         mysecret: op.Read&{
               kind: "Secret"
               metadata: name : "db-conn"
         }
        myhost: op.DecodeBase64 &{
            decodeValue: mysecret.value.DB_PUBLIC_HOST
        }
        status: myhost.value
    customStatus: |-
      ...
```

The status of velaql should be passed into the `context.velaql`.

Then users can use output/input of component like:

```
apiVersion: core.oam.dev/v1beta1
kind: Application
metadata:
  name: webapp
spec:
  components:
    - name: sample-db
      type: alibaba-rds
      properties:
        instance_name: sample-db
        account_name: oamtest
        password: U34rfwefwefffaked
        writeConnectionSecretToRef:
          name: db-conn
     outputs:
        # the output is the mysql service address
        - name: myhost
          valueFrom: context.velaql
    - name: app
      type: helm      
      inputs:
        # set the host to mysql service address
        - from: myhost
          parameterKey: properties.values.datasource.host
      properties:
        repoType: "helm"
        url: "xxx"
        chart: "app"
        version: "1.0.0"
```

**Describe alternatives you've considered**

**Additional context**

**Comment by barnettZQG** [comment]
This is a simple proposal, we could combine with configuration management requirements

```
apiVersion: core.oam.dev/v1beta1
kind: Application
metadata:
  name: webapp
spec:
  variables:
     static: 
         db_user: xxx
         db_database: xxx
         db_password: xxx
     discover:
        secret:
            name: xxx
            namespace: xxx
        nacos:
            url: xxx
            dataId: xxx
            group: xxx
  components:
    - name: sample-db
      type: alibaba-rds
      outputs:
        # the output is the mysql service address, endpoints are default provide and do not need to custom the definition.
        - name: myhost
          valueFrom: context.status.endpoints[0]
      properties:
        account_name: context.variables.db_user
    - name: app
      type: helm
      dependsOn: ["sample-db"]
      inputs:
        - from: myhost
          parameterKey: properties.values.datasource.host
      properties:
        repoType: "helm"
        url: "xxx"
        chart: "app"
        version: "1.0.0"
        values:
          datasource:
            user: context.variables.db_user
```

**Comment by chivalryq** [comment]
## Final target

passing data from a cloud resource components to other components

## Background

1. We are using `deploy` workflow step to manage all resource dispatching. At least we hope to use it in all scenairos. For now we are using both `apply-component` and `deploy`.

| Name                 |    deploy    |  apply-component  |
|-----------------|:------------:|:--------------:|
| multicluster         |      ✅       |      |
| override/replication |      ✅       |   |
| input/output         |              | ✅ |
   
2. `deploy` step's behavior of dispatch cloud-resource is not defined.

## Design

### Deploy Input/Output

Enable `deploy` step to pass data between the components it dispatches. In one deploy step, for combinations like `Cluster/Resource`, controller will guarantee that all inputs has been filled before dispatching it.

Example: "c1/configmap" will wait until "c1/svc" is dispatched and IO key endpoint is filled.

```shell
apiVersion: core.oam.dev/v1beta1
kind: Application
metadata:
  name: deploy-input-output
spec:
  components:
    - type: raw
      name: svc
      properties:
        apiVersion: v1
        kind: Service
        metadata:
          name: svc
        spec:
          selector:
            key: value
      outputs:
        - valueFrom: status.endpoint
          name: endpoint
    - type: raw
      name: configmap
      properties:
        apiVersion: v1
        kind: ConfigMap
        metadata:
          name: cm
      inputs:
        - from: endpoint
          parameterKey: data.key
  policies:
    - type: topology
      name: topo
      properties:
        clusters: ["c1", "c2"]
  workflow:
    step:
      - type: deploy
        name: deploy
        properties:
          policies: ["topo"]
```

### Deploy behavior of cloud resource

**TL;DR**

Keep Configuration CR in control plane. Dispatch generated secret to managed clusters.

**details**

A quick review about how `deploy` is executed.

```
read selected policies -> topology -> override -> replicate -> applyComponents
```

Configuration-based components will be normally processed before `applyComponents`. In `applyComponents`, deploy step will 

1. redirect configuration to local cluster.
2. dispatching generated secret which contains some connection info to specific cluster.

As a result, we will got one cloud resource and several secret in different clusters.

Note: in the first phase when combining `deploy` and cloud resource , we can disable `override` and `replication` policy for simplicity.

### How to combine deploying cloud resource and input/output

1. dispatch secrets to managed cluster 
2. set output to secret's name in Configuration status with key `secretName`
2. set input to use `secretName` and webservice component in managed cluster can read corresponding secret.

**Comment by wonderflow** [comment]
Since we have supported deploy step with data passing in https://github.com/kubevela/kubevela/pull/5161

What's the block on this issue now? @chivalryq

**Comment by chivalryq** [comment]
Here's some illustration of what we are meeting and the possible ways to solve this problem.

### Possible solutions

Solutions can be divided into two kinds:

1. **Pass Secret**: The secret that contains connection info is created at managed cluster. Workload use the Secret name and key in Secret. This is solved.
2. **Pass Value**: The secret is not created in managed cluster. The required connection info is directly written in workload spec (e.g. env). This is the issue mainly cares. Unfortunately there's more problems in this way.

### Problems

<img width="1261" alt="image" src="https://user-images.githubusercontent.com/47812250/218315381-84fa2f96-ed55-44b5-987b-65f21b469912.png">

As displayed in mind map above, the main problem to **pass Data** is read ing data from either Configuration or Secret it created. Here is an image that tells relationship between Application, Configuration and Secret, which helps to understand 2.1 and 2.2.1

<img width="1307" alt="image" src="https://user-images.githubusercontent.com/47812250/218315537-60afb56a-9152-4bb1-886f-ac949423c94e.png">

### Demo of Passing Secret

Here's a demo about passing secret from cloud resource component to other worload. In this demo application will first create a database and then create two workloads in different clusters. The both workloads will use the same DB.

```yaml
apiVersion: core.oam.dev/v1beta1
kind: Application
metadata:
  name: multi-cluster-app-with-db
spec:
  components:
    - name: db
      type: alibaba-rds
      properties:
        instance_name: favorite-links
        database_name: links
        account_name: oamtest
        password: U34rfwefwefffaked
        writeConnectionSecretToRef:
          name: db-conn
    - name: favorite-links
      type: webservice
      properties:
        image: oamdev/nodejs-mysql-links:v0.0.1
        port: 4000
      traits:
        - type: service-binding
          properties:
            envMappings:
              DATABASE_HOST:
                secret: db-conn
                key: DB_PUBLIC_HOST
              DATABASE_NAME:
                secret: db-conn
                key: DATABASE_NAME
              DATABASE_USER:
                secret: db-conn
                key: DB_USER
              DATABASE_PASSWORD:
                secret: db-conn
                key: DB_PASSWORD
  policies:
    - name: worker
      type: topology
      properties:
        clusters:
          - cluster-worker
    - name: all-cluster
      type: topology
      properties:
        clusters:
          - cluster-worker
          - local
    - name: links-comp
      type: override
      properties:
        selector:
          - favorite-links
  workflow:
    steps:
      - name: create-rds
        type: apply-component
        properties:
          component: db
      - name: read-secret
        type: read-object
        properties:
          apiVersion: v1
          kind: Secret
          name: db-conn
        outputs:
          - name: db-host
            valueFrom: |
              import "encoding/base64"
              base64.Decode(null, output.value.data.DB_PUBLIC_HOST)
          - name: db-user
            valueFrom: |
              import "encoding/base64"
              base64.Decode(null, output.value.data.DB_USER)
          - name: db-password
            valueFrom: |
              import "encoding/base64"
              base64.Decode(null, output.value.data.DB_PASSWORD)
          - name: db-name
            valueFrom: |
              import "encoding/base64"
              base64.Decode(null, output.value.data.DATABASE_NAME)
      - name: export-secret
        type: export-data
        properties:
          name: db-conn
          namespace: default
          kind: Secret
          topology: worker
        inputs:
          - from: db-host
            parameterKey: data.DB_PUBLIC_HOST
          - from: db-user
            parameterKey: data.DB_USER
          - from: db-password
            parameterKey: data.DB_PASSWORD
          - from: db-name
            parameterKey: data.DATABASE_NAME

      - name: deploy-web
        type: deploy
        properties:
          policies: [ "all-cluster", "links-comp" ]
```

### Discuss About Passing Value

Let's see what we have solved:

|     | Local | Sub-Cluster
|  ----  | ----  | ----  |
|  Pass Secret  | ✅ | ✅ |
| Pass Value  | ✅ |  |

To fill the last grid of table, we have to solve either 2.1 or 2.2.

- To solve 2.1 we need to refactor the Configuration rendering logic and unify it to CUE.
- To solve 2.2, 2.2.3 is an easier way and need less effort.

**Comment by wonderflow** [comment]
I'm going to close this issue as most of the cases have been supported, we'll raise new issues if we found use cases that not meet. https://github.com/kubevela/kubevela/pull/5489
